### DISCUSSION TITLE
accessors for arrays (and for anything in fact) in forms

### DISCUSSION THREAD

**Discussion by epertinez** [filamentphp/filament#6466] [Help] [answered]
Hi, I have what I assume is a common case and have been hours and cannot find a way to solve it.

My customer's table has a field with a serialized array (as in serialize($array) ). I need to edit those fields and, as simple as it seems, I cannot.
I tried to create a Muttator over the field in the Model, so I json it with get (json_encode(unserialize($value))) and serialize with set (serialize(json_decode($value)), but it does not work. The getter seems to be called twice but the KeyValue field acts crazy.

If I access the values one by one with dot quotation, field.someElement it saves correctly (calls the setter) but never calls the getter. As if the hydrating part is not using the getters.

I am stocked with this that I think should be easy and quite common (not everything has turned to json yet). 

I tried also to have a virtual field in the Model that gets the jsoned version and sets the serialized one, with the same result: setter is being called, but getter not.

Don't know even if it is possible, yet I think it should be as it should be common not only for arrays but normal fields...

Another thing I tried with the same result: I created a virtual field in the model... fieldsomeElement, that gets its value from the array. Same result. the getter never gets called.

Anybody knows how to resolve it?
At this moment any dirty trick is acceptable. Maybe after hydrating or something....

Any help will be appreciated...

**Comment by epertinez** [comment]
Just minutes after writing it i found the solution.
The point here is that you don't need a json object to give to Forms\Components\KeyValue::make('serialized_field') but an array.

KeyValue works with json ONLY if you cast the field to json in your model 
```
  protected $casts = [
        //...
        'jsoned_field' =>'json'
    ];
```
So I guess that it decodes the json to an array and goes on from there.

BUT, if, as in my case, you don't have a json BUT a serialized field, then, the trick is to use a mutator that returns THE ARRAY to KeyValue, not the json.

So... in your model you have to have something like
```
protected function serializedField(): Attribute
    {
        return Attribute::make(
            get: fn (?string $value) => @unserialize($value),
            set: fn (array $value) => serialize($value),
        );
    }
```

and in your Resource you can call then...
`
Forms\Components\KeyValue::make('serialized_field''),
`
to tour form schema and everything works as expected.

This does not explain why all my intents to use a virtual field getJsonedFieldAttribute()/setJsonedFieldAttribute($value) never worked because the getter  was never called. That, I understand, should be revised, because it makes no sense. I think.

**Comment by danharrin** [comment]
Ive never see anyone serialising arrays instead of just using JSON (array cast)... we fully support JSON but serialisation would require it to be cast to an array first so its usable.

**Reply by epertinez** [reply]
I (and my customer's code) am older than you.
:-D

**Reply by epertinez** [reply]
Two things:
a) Using a "virtual" field did not work as the getter was never called. Don't know exactly why, but I assume it should.
b) I think there should be a reference in the documentation explaining that you can only use json IF the field is saved in json in the database (and/or properly casted in your model). And there should be a reference in the documentation explaining that you can mutate your field so it returns an array instead of a json and that this should work. serialize is not the only way to save an unidimensional array in a string field. I imagine lots of different cases (comma separated pairs, ...) where a mutator could do the trick to use a KeyValue field.

**Reply by danharrin** [reply, marked as answer]
Please try using a cast instead of a "virtual field" (accessor). I don't think accessors get used when we `toArray()` a model to fill the form.
